lib/logstorage/storage.go

```diff
@@ -141,6 +141,22 @@ type Storage struct {
 	//
 	// It reduces the load on persistent storage during querying by _stream:{...} filter.
 	filterStreamCache *cache
+
+	// minKeptDay is set ONLY after the disk-space watcher removes recent partitions.
+	// It marks the earliest day still kept; older days are rejected during ingestion
+	// to avoid re-creating partitions just purged for space.
+	minKeptDay atomic.Int64
+}
+
+// setMinKeptDayFromPartitions updates minKeptDay based on the current list of partitions.
+// It must be called while s.partitions reflects the up-to-date state. The caller may hold
+// or not hold partitionsLock depending on the context — the function itself doesn't lock.
```

Comment:
The unclear semantics regarding the requirements on the partitionLock doesn't look like a maintainable and clear solution. It increases chances of subtle data race bugs in the future.